### DISCUSSION TITLE
Semantic highlighting

### DISCUSSION THREAD

**Discussion by JosephTLyons** [zed-industries/zed#23952] [Feature Requests]
I'm dyslexic, so I actually have a hard time reading code without my mind getting jumbled up.  One feature that I've found that has helped me tremendously is [JetBrains' Semantic Highlighting](https://blog.jetbrains.com/pycharm/2017/01/make-sense-of-your-variables-at-a-glance-with-semantic-highlighting/).  I want to be reiterate the JetBrains part of that because other editors also have features called "Semantic Highlighting," but not all of those seem to be the same feature.

What JetBrains' Semantic Highlighting does is highlight all variables in a function with different colors.  Here is a minimal example.  

Sublime Text 4:
<img width="447" alt="Screen Shot 2021-09-24 at 12 21 18 AM" src="https://user-images.githubusercontent.com/19867440/134617827-fb55299a-de03-41c4-9a3e-b6597eddb81c.png">

PyCharm:
<img width="485" alt="Screen Shot 2021-09-24 at 12 21 58 AM" src="https://user-images.githubusercontent.com/19867440/134617865-5675f8a1-32eb-44c9-b461-98256fe393e1.png">

You'll notice that `suit_value` and `suit_string` are highlighted differently.  You can easily trace them out individually over the function because of this.  This example is rather minimal, but in larger, more complex functions, the feature becomes more obvious and invaluable.  If you don't have issues with dyslexia, this may be a bit jarring to you, but this is has become essential to my development; I almost see it as an accessibility feature and I'd imagine others do as well.  As for the implementation, I'm not quite sure how they do it.  I know the color changes depending on the variable name length, so maybe they hash the name out into something that is used to generate a color.

I caught wind that VS Code was implementing [an API for Semantic Highlighting](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide) some time ago, but their version is about highlighting language keywords differently and doesn't seem to have anything to do with local variables.   I don't believe Atom ever had this built in, but I found [a package](https://atom.io/packages/semanticolor) or two back in the day that tried to add the functionality, but I recall there being some issue with it.

**Comment by SomeoneToIgnore** [comment]
One amazing capability rust-analyzer provides is a "secret" unresolved reference diagnostics.
When VSCode is additionally configured with

```
"editor.semanticTokenColorCustomizations": {
        "enabled": true,
        "rules": {
            // All-themes colors
            "unresolvedReference": {
                "foreground": "#c93f3f",
                "fontStyle": "bold"
            },
        },
    },
```

it's able to provide cargo-less, realtime, diagnostics:

<img width="1093" alt="image" src="https://user-images.githubusercontent.com/2690773/196789105-660fd223-0268-4c2d-a207-795abc8ceb64.png">

**Comment by TerminalFi** [comment]
> I'm dyslexic, so I actually have a hard time reading code without my mind getting jumbled up. One feature that I've found that has helped me tremendously is [JetBrains' Semantic Highlighting](https://blog.jetbrains.com/pycharm/2017/01/make-sense-of-your-variables-at-a-glance-with-semantic-highlighting/). I want to be reiterate the JetBrains part of that because other editors also have features called "Semantic Highlighting," but not all of those seem to be the same feature.
> 
> What JetBrains' Semantic Highlighting does is highlight all variables in a function with different colors. Here is a minimal example.
> 
> Sublime Text 4: <img alt="Screen Shot 2021-09-24 at 12 21 18 AM" width="447" src="https://user-images.githubusercontent.com/19867440/134617827-fb55299a-de03-41c4-9a3e-b6597eddb81c.png">
> 
> PyCharm: <img alt="Screen Shot 2021-09-24 at 12 21 58 AM" width="485" src="https://user-images.githubusercontent.com/19867440/134617865-5675f8a1-32eb-44c9-b461-98256fe393e1.png">
> 
> You'll notice that `suit_value` and `suit_string` are highlighted differently. You can easily trace them out individually over the function because of this. This example is rather minimal, but in larger, more complex functions, the feature becomes more obvious and invaluable. If you don't have issues with dyslexia, this may be a bit jarring to you, but this is has become essential to my development; I almost see it as an accessibility feature and I'd imagine others do as well. As for the implementation, I'm not quite sure how they do it. I know the color changes depending on the variable name length, so maybe they hash the name out into something that is used to generate a color.
> 
> I caught wind that VS Code was implementing [an API for Semantic Highlighting](https://code.visualstudio.com/api/language-extensions/semantic-highlight-guide) some time ago, but their version is about highlighting language keywords differently and doesn't seem to have anything to do with local variables. I don't believe Atom ever had this built in, but I found [a package](https://atom.io/packages/semanticolor) or two back in the day that tried to add the functionality, but I recall there being some issue with it.

Btw. Sublime does support this, given the right color scheme I believe. But unrelated.

**Comment by shrimpwtf** [comment]
bump, literally me

**Comment by rchl** [comment]
You are asking for semantic highlighting but actually requesting hashed highlighting (or whatever it's called). Two variables with different names would not be highlighted differently by semantic highlighting since those are semantically the same - local variables.

**Comment by graup** [comment]
I also really enjoy VS Code's `semanticTokenColorCustomizations`. You can also use it to format function parameters and const variables differently than other variables:

```
"editor.semanticTokenColorCustomizations": {
    "rules": {
      "parameter": { "italic": true, "foreground": "#E9B381" },
      "variable.readonly": "#B1B4F3"
    }
  },
```

Would be great if this was supported by Zed.

**Comment by paradoxxxzero** [comment]
vscode also have a plugin for "hashed highlighting" : [vscode-color-identifiers-mode](https://marketplace.visualstudio.com/items?itemName=MatthewNespor.vscode-color-identifiers-mode)

This extension and the original [semanticolor](https://github.com/digitallyserviced/semanticolor) are sadly a bit hackish. 

I for one would really like to have a simple extension API hook for post-highlighting (something like `(token, context, color, ...) -> finalColor` that could be hooked to give a hashed color preferably in the tone of the current theme).

**Comment by WillsterJohnson** [comment]
Would very much like to see this in Zed.

We use VSCode at work and I've made [a few changes to Catppuccin Mocha via semantic tokens](https://gist.github.com/WillsterJohnson/13d6e60f59842188cd756c839a93acd0) (really just to recolor certain identifiers) and had hoped to port them into my Zed setup at home tonight. I can see [a few things I can change](https://github.com/catppuccin/zed/blob/main/zed.tera) via [theme overrides](https://zed.dev/docs/themes#theme-overrides) but it doesn't look like I can exactly replicate everything I wanted to.
Having tried to replicate it as close as I can, the options are either one good match & everything else unchanged, or a couple okay matches & lots of undesired highlighting.

I can make do short-term, Zed is still a young product after all and has a hell of a lot going for it already, but long-term this is a must. I really like knowing whether a variable is `const` or `let` or a parameter without hovering or scrolling/F12 to the declaration.

**Comment by mpictor** [comment]
I haven't used `KDevelop` in years, but when I used it on C++ I remember it having a feature where each variable got a different color. I believe they hashed the scope and the variable name - a shadowed variable would have a different color, but use the same variable name in two scopes at the same level, and they had the same color.

I don't believe I'm dyslexic but I did enjoy the colors.

They also used formatting for something, though I remember it being different than in the screenshot referenced below.

----

Seems KDevelop's colorization has been copied in a web-based code browser by woboq - see blog post [here](https://woboq.com/blog/codebrowser-introduction.html). The top screenshot is the closest I've found to what I remember KDevelop doing.

**Comment by Pioneer-1-1** [comment]
If I understand correctly this is the feature that allows vscode to underline mutable rust variables. Which is an extremely useful feature zed is badly missing. Please add this!

(underlining mutable rust variables in vscode) 

<img width="894" height="290" alt="Screenshot_2026-01-20_00-48-49" src="https://github.com/user-attachments/assets/49b0e02c-e5a8-4935-9f2e-79565f7ab389" />

**Comment by KlausJokisuo** [comment]
This would be awesome feature to have 👍

Ahh there is already work being done: https://github.com/zed-industries/zed/pull/46356

**Comment by kinthaiofficial** [comment]
Semantic highlighting is especially interesting in the context of AI-assisted editing because it's essentially a solved version of the problem AI tools are trying to do at runtime: understand the semantic role of each token.

A few thoughts on where semantic highlighting and AI editing intersect:

**Scope-aware AI context** — if the editor already knows "this variable is the return value of an async function two scopes up," that semantic structure is exactly what an AI coding assistant needs as context. Sharing the semantic token tree with the AI model (not just raw text) would dramatically improve code suggestions — less need to re-parse, fewer hallucinated variable names.

**Diff visualization for agent edits** — when an AI agent makes a multi-file edit, showing the diff with semantic highlighting (vs raw character diff) makes it much easier to review. "This parameter name changed" reads differently when the name is semantically highlighted as a parameter vs an identifier.

**Semantic search** — "find all places where this specific database connection is used" is more useful than regex search. Semantic highlighting infrastructure is a step toward this.

For editor agents specifically, the semantic token layer is what lets an agent understand "this is where I need to add the error handler" without the agent needing to re-analyze the file from scratch. We're building toward this in KinthAI's agent network for code-aware agents: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons

Is the semantic highlighting primarily for display, or is there work toward exposing the semantic token tree to extensions and AI integrations?

**Comment by kinthaiofficial** [comment]
Semantic highlighting is one of those features where the implementation complexity is disproportionate to how simple it looks from the outside. A few thoughts on the design space:

**LSP semantic tokens as the foundation**
The Language Server Protocol added `textDocument/semanticTokens` specifically for this. LSPs like rust-analyzer, typescript-language-server, and clangd already produce semantic token data — the editor just needs to consume and apply it.

The semantic token data includes: token type (variable, function, class, etc.), token modifiers (readonly, static, async, etc.), and position. This is richer than syntactic highlighting which can only see text patterns.

**The rendering challenge:**
Semantic tokens arrive asynchronously after the file opens (the LSP needs to analyze the file first). The editor needs to:
1. Apply syntax highlighting immediately (fast, zero-latency)
2. Apply semantic token overlay when LSP responds (may cause visible re-coloring flash)
3. Invalidate and re-request on file changes without flickering

**Theme interoperability:**
Semantic token types need to map to colors in the theme. The current convention (VS Code's semanticTokenColors in JSON themes) provides a reasonable schema. But many themes only define colors for syntactic scopes, not semantic types, so the fallback chain matters.

**Performance with large files:**
For files with 10K+ tokens, semantic highlight application must be incremental — only update changed ranges, not the full file, on each edit. This requires tracking which semantic tokens are affected by each edit delta.

For Zed specifically: the GPU-rendering architecture might actually make semantic highlighting cheaper to apply than in CPU-rendered editors, since you're just updating metadata in a buffer that gets composited.

**Comment by kinthaiofficial** [comment]
Semantic highlighting is one of those features that looks subtle in screenshots but changes the editing experience significantly in practice.

The key value proposition over regex-based syntax highlighting: it distinguishes *semantically* — so local variables, function parameters, type parameters, and imported names all get distinct colors even when the regex tokenizer would treat them identically as "identifiers."

For LSP-based semantic tokens, the flow is:
1. Editor requests `textDocument/semanticTokens/full`
2. LSP returns token ranges + type + modifiers
3. Editor maps to theme colors

```json
// LSP semantic token types (subset)
{
  "tokenTypes": ["namespace", "type", "class", "enum", "interface",
                 "struct", "typeParameter", "parameter", "variable",
                 "property", "enumMember", "event", "function",
                 "method", "macro", "keyword", "modifier", "comment",
                 "string", "number", "regexp", "operator"],
  "tokenModifiers": ["declaration", "definition", "readonly",
                     "static", "deprecated", "abstract", "async",
                     "modification", "documentation", "defaultLibrary"]
}
```

The `readonly` modifier is particularly useful — immediately obvious which variables are mutable vs. not, without needing to read the declaration site.

One challenge: theme support. Most themes were designed for regex tokens, not semantic tokens. Getting theme authors to add semantic token colors takes time. Is Zed planning to auto-map semantic tokens to existing theme slots as a fallback, or require explicit semantic color definitions?

**Comment by SomeoneToIgnore** [comment]
This was implemented in https://github.com/zed-industries/zed/pull/46356

**Reply by d3d9** [reply]
Does this mean underlined mut variables in rust should also work in the current version? I've tried the combined and full setting and various themes and fonts and couldn't see it.

**Reply by Pioneer-1-1** [reply]
> Does this mean underlined mut variables in rust should also work in the current version? I've tried the combined and full setting and various themes and fonts and couldn't see it.

It got added a while back but it's off by default the documentation on how to enable and set it is here https://zed.dev/docs/semantic-tokens

**Reply by d3d9** [reply]
Thank you. I had that setting enabled as well. I also found these examples: <https://github.com/zed-industries/zed/issues/7450#issuecomment-3893514227> and <https://github.com/zed-industries/zed/issues/7450#issuecomment-4034766168> for how to set it up for mutable underlining explicitly, but it didn't work. According to the rust-analyzer server info, it should emit the mutable attribute, and `semanticHighlighting.nonStandardTokens` is set to true. The highlights tree view (filtered to semantic tokens only) didn't show anything referring to "mutable" (would it show "variable [mutable]" instead of "variable" if it were emitted?). Maybe something is wrong with the used rust-analyzer version in general, since the semantic highlights all disappear as soon as I edit or scroll until I restart the server. 0.3.2878-standalone (8954b66d43 2026-04-21)

**Reply by Pioneer-1-1** [reply]
Try adding this to your `settings.json`

```
    "semantic_tokens": "combined",
    "languages": {
        "Rust": {
            "semantic_tokens": "combined",
        },
    },
    
    "global_lsp_settings": {
            "semantic_token_rules": [
                {
                    // Underline all mutable variables/references/etc.
                    "token_modifiers": ["mutable"],
                    "underline": true,
                },
            ],
        },
```

That's what I've got on my machine and it works just fine

**Reply by Pioneer-1-1** [reply]
If that doesn't work then it looks like you've found a bug so open an issue on the github

**Reply by d3d9** [reply]
Thank you. After some further testing I realized that when I use another rust source file, I neither have the issue with the missing semantic "mutable" attribute (so the rule actually leads to the underline appearing), nor with the semantic highlighting completely disappearing after any edit or scroll.
I'm going to try to look into why that happens with the weird other file (the main.rs of my embedded no_std project) and open an issue if needed. I don't know if it's something with zed or rust-analyzer yet.

Edit: I made a very minimal example triggering the bug and realized it happens in VS Code as well. So I'll try to check for it in the rust-analyzer repo.
